README.md
# ESP8266 FastLED IoT Webserver

<a href="https://github.com/NimmLor/esp8266-fastled-iot-webserver/graphs/contributors" alt="Contributors"><img src="https://img.shields.io/github/contributors/NimmLor/esp8266-fastled-iot-webserver" /></a> <a href="https://github.com/NimmLor/esp8266-fastled-iot-webserver/releases" alt="Downloads"><img src="https://img.shields.io/github/downloads/NimmLor/esp8266-fastled-iot-webserver/total"/></a> <a href="https://github.com/NimmLor/esp8266-fastled-iot-webserver/commits/master" alt="Downloads"><img src="https://img.shields.io/github/commits-since/NimmLor/esp8266-fastled-iot-webserver/latest?include_prereleases" /></a> <a href="https://github.com/NimmLor/esp8266-fastled-iot-webserver/releases" alt="Downloads"><img src="https://img.shields.io/github/downloads-pre/NimmLor/esp8266-fastled-iot-webserver/latest/total?color=%09%23228B22&label=download%40latest-version" /></a>


This is a fork of [jasoncoon's esp8266 fastled webserver](https://github.com/jasoncoon/esp8266-fastled-webserver) that was adapted to control the colors of my  [LED-Projects](https://www.thingiverse.com/Surrbradl08/designs).

![Webinterface](web_ui.jpg?raw=true)

The web app is stored in SPIFFS (on-board flash memory) and has around 400kb. All stylsheets/js used are also stored in the spiffs, so no internet connection is required to view the webpage correctly.

**New:** Windows application to sync the LED devices with music. Completely optional but it's awesome. But it's just the first prototype and very experimental. Software and docs still work in progress.

https://github.com/NimmLor/IoT-Audio-Visualization-Center 

Can be downloaded [here](https://github.com/NimmLor/IoT-Audio-Visualization-Center/releases).

![screenshot_alpha](screenshot_alpha.jpg?raw=true)

https://github.com/NimmLor/IoT-Audio-Visualization-Center



## What happened here?

All of my recent project were merged into one including most of the features. A new and optional Windows application allows to sync the LEDs to a Windows audio source of your choice.



### Supported Devices

**Generic LED-Strip**, just a regular LED-Strip without special hardware

  * Easiest: 5V WS2812B LED-Strip:		https://s.click.aliexpress.com/e/_dZ1hCJ7
  * (Long Ranges) 12V WS2811 LED-Strip:	https://s.click.aliexpress.com/e/_d7Ehe3L
  * (High-Speed) 5V SK9822 LED-Strip:		https://s.click.aliexpress.com/e/_d8pzc89
  * (Expensive) 5V APA102 LED-Strip:		https://s.click.aliexpress.com/e/_Bf9wVZUD
  * (Flexible) 5V WS2812 S LED-Strip:		https://s.click.aliexpress.com/e/_d6XxPOH
  * Wemos D1 Mini:						https://s.click.aliexpress.com/e/_dTVGMGl
  * 5V Power Supply:						https://s.click.aliexpress.com/e/_dY5zCWt
  * Solderless LED-Connector:				https://s.click.aliexpress.com/e/_dV4rsjF
  * 3D-Printed Wemos-D1 case:				https://www.thingiverse.com/thing:3544576

**LED-Matrix**, with a flexible LED-Matrix you can display the audio like a Audio Visualizer
* Flexible WS2812 Matrix:				https://s.click.aliexpress.com/e/_d84R5kp
	
* Wemos D1 Mini:						https://s.click.aliexpress.com/e/_dTVGMGl
* 5V Power Supply:						https://s.click.aliexpress.com/e/_dY5zCWt
	

**3D-Printed 7-Segment Clock**, display the time, syncs with a ntp server of your choice

* unfortunately the "thing's" description isn't updated yet to the new standalone system
	
* Project link, small version:			https://www.thingiverse.com/thing:3117494
* Project link, large version:			https://www.thingiverse.com/thing:2968056

**3D-Printed Desk Lamp**, a lamp that reacts to sound for your desk

* Project link, twisted version:		https://www.thingiverse.com/thing:4129249
* Project link, round version:			https://www.thingiverse.com/thing:3676533

**3D-Printed Nanoleafs**, a Nanoleaf clone that can be made for cheap

* Project link:													https://www.thingiverse.com/thing:3354082
	

**3D-Printed Animated RGB Logos**, a small 3D-Printed logo that lights up with style

* Project link, Twenty-One-Pilots Logo:		https://www.thingiverse.com/thing:3523487
* Project link, Thingiverse Logo:			https://www.thingiverse.com/thing:3531086

**(soon) 3D-Printed Infinity Mirror**, a DIY infinity mirror inspired by Adafruit

* Project link, Sound Reactive Infinity Mirror:		https://www.thingiverse.com/thing:4461070

**3D-Printed IoT Bottle Lighting Pad**, a simple but smart lighting for any bottle you want.

* Project link, IoT Bottle Lighting Pad:		https://www.thingiverse.com/thing:4461313



## Changelog

### 07.05.2020, Major Code rewrite and merge of projects, audio visualization

- **New features:**
  - **Audio Visualization with a Windows Desktop Application (C#, WPF)** [here](https://github.com/NimmLor/IoT-Audio-Visualization-Center)
  - multicast DNS by @WarDrake
  - OTA Support
  - MQTT/Homeassistant integration by @WarDrake
  - Serial Ambilight for usage behind a TV
  - Support of Desk Lamp, 7-Segment Clock, Animated RGB Logos, Generic LED-Strip
  - WebUI fits now on 1MB devices (esp-01)
  - Dark mode for WebUI

### 01.02.2020, Native Alexa Update

- **NodeRED** part is now **DEPRECATED**
- The Nanoleaf Replica allows now for **NATIVE** Alexa support without the need of an extra Raspberry Pi. When added to the Smart Home devices in the Alexa app, the nanoleafs will appear as Phillips Hue devices.
- Added Strobe Pattern
- Added Sound Reactive support
- Some code cleanup and new parameters to configure
- New step by step installation instructions ([Software_Installation.md](Software_Installation.md))



### 24.02.2019, NodeRED Update (Deprecated)

- Node-RED integration was added
- Alexa support via NodeRED


Installation
--------
**FOR INSTALLATION REFER TO THE [Software_Installation.md](Software_Installation.md)**



Features
--------
* Turn the LEDs on and off
* Appear as an **ALEXA SMART HOME DEVICE**
* **Sound Reactive Mode**
* Adjust the brightness, color and patterns
* Play over 30+ patterns in Autoplay
* ESP32 support (experimental)



## Technical

Patterns are requested by the app from the ESP8266, so as new patterns are added, they're automatically listed in the app.

The web app is stored in SPIFFS (on-board flash memory).

The web app is a single page app that uses [jQuery](https://jquery.com) and [Bootstrap](http://getbootstrap.com).  It has buttons for On/Off, a slider for brightness, a pattern selector, and a color picker (using [jQuery MiniColors](http://labs.abeautifulsite.net/jquery-minicolors)).  Event handlers for the controls are wired up, so you don't have to click a 'Send' button after making changes.  The brightness slider and the color picker use a delayed event handler, to prevent from flooding the ESP8266 web server with too many requests too quickly.

The only drawback to SPIFFS that I've found so far is uploading the files can be extremely slow, requiring several minutes, sometimes regardless of how large the files are.  It can be so slow that I've been just developing the web app and debugging locally on my desktop (with a hard-coded IP for the ESP8266), before uploading to SPIFFS and testing on the ESP8266.



### Alexa

The code has an optional feature to be able to control the lamp via Alexa on any Amazon Echo device. For setup instructions refer to [Software_Installation.md](Software_Installation.md) document.



### Compression

The web app files can be gzip compressed before uploading to SPIFFS by running the following command:

`gzip -r data/`

The ESP8266WebServer will automatically serve any .gz file.  The file index.htm.gz will get served as index.htm, with the content-encoding header set to gzip, so the browser knows to decompress it.  The ESP8266WebServer doesn't seem to like the Glyphicon fonts gzipped, though, so I decompress them with this command:

`gunzip -r data/fonts/`

### REST Web services

The firmware implements basic [RESTful web services](https://en.wikipedia.org/wiki/Representational_state_transfer) using the ESP8266WebServer library.  Current values are requested with HTTP GETs, and values are set with POSTs using query string parameters.

#### URL Endpoints
|Endpoint|Method|Keys|Attribute|Description|
|--------|------|---|---------|-----------|
/config.json|GET|N/A|N/A|returns device config in JSON format
/reboot|GET/POST|N/A|N/A|reboots the device
/power|GET/POST|value|on, 1, 0, off, toggle|switch state of LEDs
/speed|GET/POST|value|0-255|set speed of pattern/effect
/brightness|GET/POST|value|0-255|set LED brightness
/solidColor|GET/POST|r,g,b|0-255 each|set RGB values for solid color
/hue|GET/POST|value|0-255|set hue: [FastLED HUE](https://github.com/FastLED/FastLED/wiki/FastLED-HSV-Colors)
/saturation|GET/POST|value|0-255|set saturation
/patternName|GET/POST|value|pattern name|define a certain pattern, pattern names can be retreived via `/config.json` endpoint
/paletteName|GET/POST|value|palette name|defien a certain palette which is used by some effects, palette names can be retreived via `/config.json` endpoint
/autoplay|GET/POST|value|on, 1, 0, off, toggle|switch autoplay LED patterns on or off
/autoplayDuration|GET/POST|value|1-255|set time in seconds each pattern should be autoplayed
/twinkleDensity|GET/POST|value|0-255|set density of some twinkle effects
/cooling|GET/POST|value|0-255|set sparking intensity for `Fire` and `Water` pattern
/sparking|GET/POST|value|0-255|set cooling intensity for `Fire` and `Water` pattern
/reset|POST|type|wifi, all|wipe config just for `wifi` or `all` settings incl. wifi
/settings|POST|ssid|WiFi SSID|set a new WiFi SSID
/settings|POST|password|WiFi password|set a new WiFi password
/settings|POST|hostname|device hostname|set a new device hostname
/settings|POST|mqtt-enabled\*|1, 0|enable/disable MQTT support
/settings|POST|mqtt-hostname\*|hostname|set the hostname of the MQTT server to connect to
/settings|POST|mqtt-port\*|1-65535|set the MQTT port to connect to
/settings|POST|mqtt-user\*|a username|set the MQTT username to be used during MQTT connection
/settings|POST|mqtt-password\*|a password|set the MQTT password to be used during MQTT connection
/settings|POST|mqtt-topic\*|a topic|set the MQTT topic to be used during MQTT connection
/settings|POST|mqtt-device-name\*|device hostname|set the MQTT topic to be used during MQTT connection

>\* MQTT support has to be enabled during firmware build (disabled by default). All MQTT settings (except mqtt-enable) trigger a reboot

### MQTT Syntax

The MQTT integration by default sets up a light domain autodiscovery for home assistant, should you want to send MQTT commands from a different system or manually the syntax is as follows

**Topic :** MQTT_TOPIC / MQTT_TOPIC_SET  
example : "homeassistant/light/nanoleafs/set" in the case of the nanoleafs if you use default values, adjust for any changes you make to the previous two configuration variables  

**payload :** this is a json formatted string with parameters, you can send one or multiple parameters, please note that to change most of them the light must be on, so sending an on command on every request other than off is a good idea.  

**commands :**
|Key         |Value            |
|------------|-----------------|
|state       |on, off, toggle
|brightness  |1 to 255
|autoplay    |on, off, toggle
|speed       |1 to 255
|effect      |Pattern name as a quoted string.
|color       |[1..255,1..255,1..255]
|hue         |1 to 255
|saturation  |1 to 255

**command examples :**
|example                   |JSON payload|
|--------------------------|------------|
|turn lights on            |`{"state": "ON"}`
|turn lights off           |`{"state": "OFF"}`
|Set brightness to 50%     |`{"state": "ON", "brightness": 127}`
|Set animation speed to 16 |`{"speed": "16"}`
|Set animation autoplay on |`{"autoplay": "ON"}`
|Set animation pattern     |`{"state": "ON", "effect": "Sinelon"}`
|Set a solid color         |`{"state": "ON", "color": {"r": 72, "g": 255, "b": 163}}`

**You can combine multiple commands in a single payload, for example:**

Set lights on, with a brightness of 50%, animation speed of 25, animation autplay off and the pride pattern.
```json
{"state": "ON", "brightness": 127, "speed": "25", "autoplay": "OFF", "effect": "Pride"}
```

Changelog.md
# Changelog

## 4.5 - 14.02.2021, Config web page, ESP32 support, Homey integration

### Features:
- adds settings page to remove secrets.h and configure Wifi, hostname and MQTT settings #94
- Homey integration #117
- calculate delay per frame dynamically to keep frame rate steady #125
- adds Serial DEBUG mode and improves DEBUG information #124
- disable certain web interface controls if not available/used by pattern #92
- adds ESP32 support (experimental) #77
- adds more controls (speed/pattern/twinkle density) to more patterns #112
- UI / CSS Rework changes the sliders #132

### Bugfixes:
- fix compile errors when MQTT support was enabled #109
- fixes problem with UDP visualization not working #103
- fixes issue with unsaved config LEDs were power off #127
- fixes duplicate open SSID after wifiManager connected #135
- fixes random 4 seconds freezes #114
- fixes an compile error if LED_DEVICE_TYPE 5 was selected #73
- fixes Wifi setup if `LED_DEVICE_TYPE 2` was selected #100
- fixes HomeAssistant not recognizing separate esp boards #148

### House keeping:
- remove half integrated simple page #128
- remove unused "webSockets" support #130
- improves handling of config data written to EEPROM #121
- general code cleanup #129

### Documentation
- adds documentation of MQTT payload options #126, #137
- adds documentation of all URL endpoints and their options #147
- improves the list of dependencies and library versions #145

## 07.05.2020, Major Code rewrite and merge of projects, audio visualization

- **Audio Visualization with a Windows Desktop Application (C#, WPF)** [here](https://github.com/NimmLor/IoT-Audio-Visualization-Center)
- multicast DNS by @WarDrake
- OTA Support
- MQTT/Homeassistant integration by @WarDrake
- Serial Ambilight for usage behind a TV
- Support of Desk Lamp, 7-Segment Clock, Animated RGB Logos, Generic LED-Strip
- WebUI fits now on 1MB devices (esp-01)
- Dark mode for WebUI

## 01.02.2020, Native Alexa Update

- **NodeRED** part is now **DEPRECATED**
- The Nanoleaf Replica allows now for **NATIVE** Alexa support without the need of an extra Raspberry Pi. When added to the Smart Home devices in the Alexa app, the nanoleafs will appear as Phillips Hue devices.
- Added Strobe Pattern
- Added Sound Reactive support
- Some code cleanup and new parameters to configure
- New step by step installation instructions ([Software_Installation.md](Software_Installation.md))

## 24.02.2019, NodeRED Update (Deprecated)

- Node-RED integration was added
- Alexa support via NodeRED


config.h

// define EEPROM settings
//  https://www.kriwanek.de/index.php/de/homeautomation/esp8266/364-eeprom-für-parameter-verwenden

typedef struct {
  uint8_t brightness;
  uint8_t currentPatternIndex;
  uint8_t red;
  uint8_t green;
  uint8_t blue;
  uint8_t power;
  uint8_t autoplay;
  uint8_t autoplayDuration;
  uint8_t currentPaletteIndex;
  uint8_t speed;
  char hostname[33];
  uint8_t MQTTEnabled;
  char MQTTHost[65];
  uint16_t MQTTPort;
  char MQTTUser[33];
  char MQTTPass[65];
  char MQTTTopic[65];
  char MQTTSetTopic[65];
  char MQTTDeviceName[33];
} configData_t;


configData_t cfg;
configData_t default_cfg;

// set to true if config has changed
bool save_config = false;


void saveConfig(bool save) {
    // Save configuration from RAM into EEPROM
    if (save == true) {
        EEPROM.begin(4095);
        EEPROM.put(0, cfg );
        delay(200);
        EEPROM.commit();
        EEPROM.end();

        save_config = false;
    }
}

void resetConfig() {

        // delete EEPROM config
        EEPROM.begin(4095);
        for (int i = 0 ; i < sizeof(cfg) ; i++) {
            EEPROM.write(i, 0);
        }
        delay(200);
        EEPROM.commit();
        EEPROM.end();

        // set to default config
        cfg = default_cfg;
        saveConfig(true);  
}

void setHostname(String new_hostname)
{
    int j = 0;
    for (int i = 0; i < new_hostname.length() && i < sizeof(cfg.hostname); i++) {
        if (new_hostname.charAt(i) == '-' or \
           (new_hostname.charAt(i) >= '0' && new_hostname.charAt(i) <= '9') or \
           (new_hostname.charAt(i) >= 'A' && new_hostname.charAt(i) <= 'Z') or \
           (new_hostname.charAt(i) >= 'a' && new_hostname.charAt(i) <= 'z')) {

            cfg.hostname[j] = new_hostname.charAt(i);
            j++;
        }
    }
    cfg.hostname[j] = '\0';
    save_config = true;
}
// EOF


data/js/app.js
// used when hosting the site on the ESP8266
var address = location.hostname;
var urlBase = "";

// used when hosting the site somewhere other than the ESP8266 (handy for testing without waiting forever to upload to SPIFFS)
// var address = "esp8266-1920f7.local";
// var urlBase = "http://" + address + "/";

var postColorTimer = {};
var postValueTimer = {};

var ignoreColorChange = false;

var patternData = {};

$(document).ready(function() {
  $("#status").html("Connecting, please wait...");

  $.get(urlBase + "config.json", function(data) {
      $("#status").html("Loading, please wait...");

      $.each(data, function(index, field) {
        if (field.type == "Number") {
          addNumberField(field);
        } else if (field.type == "Boolean") {
          addBooleanField(field);
        } else if (field.type == "Select") {
          addSelectField(field);
        } else if (field.type == "Color") {
          addColorFieldPalette(field);
          addColorFieldPicker(field);
        } else if (field.type == "Section") {
          addSectionField(field);
        } else if (field.type == "Setting") {
          handleSetting(field);
        }
        if (field.name == "pattern") {
          patternData = field;
        }
      });

      $(".minicolors").minicolors({
        theme: "bootstrap",
        changeDelay: 200,
        control: "wheel",
        format: "rgb",
        inline: true
      });

      updateDisplayedFields();

      $("#status").html("Ready");
    })
    .fail(function(errorThrown) {
      console.log("error: " + errorThrown);
    });
});

function addNumberField(field) {
  var template = $("#numberTemplate").clone();

  template.attr("id", "form-group-" + field.name);
  template.attr("data-field-type", field.type);

  var label = template.find(".control-label");
  label.attr("for", "input-" + field.name);
  var text = field.label.indexOf("Autoplay") > -1 ? field.label + ": " + field.value + " seconds" :field.label + ": " + (field.value * 100 / 255).toFixed(0) + "%"
  label.text(text);

  var input = template.find(".input");
  var slider = template.find(".slider");
  slider.attr("id", "input-" + field.name);
  if (field.min) {
    input.attr("min", field.min);
    slider.attr("min", field.min);
  }
  if (field.max) {
    input.attr("max", field.max);
    slider.attr("max", field.max);
  }
  if (field.step) {
    input.attr("step", field.step);
    slider.attr("step", field.step);
  }
  input.val(field.value);
  slider.val(field.value);

  slider.on("change mousemove", function() {
    input.val($(this).val());
  });

  slider.on("change", function() {
    var value = $(this).val();
    input.val(value);
    field.value = value;
    label.text(field.label.indexOf("Autoplay") > -1 ? field.label + ": " + field.value + " seconds" :field.label + ": " + (field.value * 100 / 255).toFixed(0) + "%");
    delayPostValue(field.name, value);
  });

  input.on("change", function() {
    var value = $(this).val();
    slider.val(value);
    field.value = value;
    delayPostValue(field.name, value);
  });

  $("#form").append(template);
}

function addBooleanField(field) {
  var template = $("#booleanTemplate").clone();

  template.attr("id", "form-group-" + field.name);
  template.attr("data-field-type", field.type);

  var label = template.find(".control-label");
  label.attr("for", "btn-group-" + field.name);
  label.text(field.label);

  var btngroup = template.find(".btn-group");
  btngroup.attr("id", "btn-group-" + field.name);

  var btnOn = template.find("#btnOn");
  var btnOff = template.find("#btnOff");

  btnOn.attr("id", "btnOn" + field.name);
  btnOff.attr("id", "btnOff" + field.name);

  btnOn.attr("class", field.value ? "btn btn-primary" : "btn btn-default");
  btnOff.attr("class", !field.value ? "btn btn-primary" : "btn btn-default");

  btnOn.click(function() {
    setBooleanFieldValue(field, btnOn, btnOff, 1)
  });
  btnOff.click(function() {
    setBooleanFieldValue(field, btnOn, btnOff, 0)
  });

  $("#form").append(template);
}

function addSelectField(field) {
  var template = $("#selectTemplate").clone();

  template.attr("id", "form-group-" + field.name);
  template.attr("data-field-type", field.type);

  var id = "input-" + field.name;

  var label = template.find(".control-label");
  label.attr("for", id);
  label.text(field.label);

  var select = template.find(".form-control");
  select.attr("id", id);

  for (var i = 0; i < field.options.length; i++) {
    var optionText = field.options[i].name;
    var option = $("<option></option>");
    option.text(optionText);
    option.attr("value", i);
    select.append(option);
  }

  select.val(field.value);

  select.change(function() {
    var value = template.find("#" + id + " option:selected").index();
    postValue(field.name, value);
    if (field.name == "pattern") {
      updateDisplayedFields(value);
    }
  });

  var previousButton = template.find(".btn-previous");
  var nextButton = template.find(".btn-next");

  previousButton.click(function() {
    var value = template.find("#" + id + " option:selected").index();
    var count = select.find("option").length;
    value--;
    if(value < 0)
      value = count - 1;
    select.val(value);
    postValue(field.name, value);
    if (field.name == "pattern") {
      updateDisplayedFields(value);
    }
  });

  nextButton.click(function() {
    var value = template.find("#" + id + " option:selected").index();
    var count = select.find("option").length;
    value++;
    if(value >= count)
      value = 0;
    select.val(value);
    postValue(field.name, value);
    if (field.name == "pattern") {
      updateDisplayedFields(value);
    }
  });

  $("#form").append(template);
}

function addColorFieldPicker(field) {
  var template = $("#colorTemplate").clone();

  template.attr("id", "form-group-" + field.name);
  template.attr("data-field-type", field.type);

  var id = "input-" + field.name;

  var input = template.find(".minicolors");
  input.attr("id", id);

  if(!field.value.startsWith("rgb("))
    field.value = "rgb(" + field.value;

  if(!field.value.endsWith(")"))
    field.value += ")";

  input.val(field.value);

  var components = rgbToComponents(field.value);

  var redInput = template.find(".color-red-input");
  var greenInput = template.find(".color-green-input");
  var blueInput = template.find(".color-blue-input");

  var redSlider = template.find(".color-red-slider");
  var greenSlider = template.find(".color-green-slider");
  var blueSlider = template.find(".color-blue-slider");

  redInput.attr("id", id + "-red");
  greenInput.attr("id", id + "-green");
  blueInput.attr("id", id + "-blue");

  redSlider.attr("id", id + "-red-slider");
  greenSlider.attr("id", id + "-green-slider");
  blueSlider.attr("id", id + "-blue-slider");

  redInput.val(components.r);
  greenInput.val(components.g);
  blueInput.val(components.b);

  redSlider.val(components.r);
  greenSlider.val(components.g);
  blueSlider.val(components.b);

  redInput.on("change", function() {
    var value = $("#" + id).val();
    var r = $(this).val();
    var components = rgbToComponents(value);
    field.value = r + "," + components.g + "," + components.b;
    $("#" + id).minicolors("value", "rgb(" + field.value + ")");
    redSlider.val(r);
  });

  greenInput.on("change", function() {
    var value = $("#" + id).val();
    var g = $(this).val();
    var components = rgbToComponents(value);
    field.value = components.r + "," + g + "," + components.b;
    $("#" + id).minicolors("value", "rgb(" + field.value + ")");
    greenSlider.val(g);
  });

  blueInput.on("change", function() {
    var value = $("#" + id).val();
    var b = $(this).val();
    var components = rgbToComponents(value);
    field.value = components.r + "," + components.g + "," + b;
    $("#" + id).minicolors("value", "rgb(" + field.value + ")");
    blueSlider.val(b);
  });

  redSlider.on("change", function() {
    var value = $("#" + id).val();
    var r = $(this).val();
    var components = rgbToComponents(value);
    field.value = r + "," + components.g + "," + components.b;
    $("#" + id).minicolors("value", "rgb(" + field.value + ")");
    redInput.val(r);
  });

  greenSlider.on("change", function() {
    var value = $("#" + id).val();
    var g = $(this).val();
    var components = rgbToComponents(value);
    field.value = components.r + "," + g + "," + components.b;
    $("#" + id).minicolors("value", "rgb(" + field.value + ")");
    greenInput.val(g);
  });

  blueSlider.on("change", function() {
    var value = $("#" + id).val();
    var b = $(this).val();
    var components = rgbToComponents(value);
    field.value = components.r + "," + components.g + "," + b;
    $("#" + id).minicolors("value", "rgb(" + field.value + ")");
    blueInput.val(b);
  });

  redSlider.on("change mousemove", function() {
    redInput.val($(this).val());
  });

  greenSlider.on("change mousemove", function() {
    greenInput.val($(this).val());
  });

  blueSlider.on("change mousemove", function() {
    blueInput.val($(this).val());
  });

  input.on("change", function() {
    if (ignoreColorChange) return;

    var value = $(this).val();
    var components = rgbToComponents(value);

    redInput.val(components.r);
    greenInput.val(components.g);
    blueInput.val(components.b);

    redSlider.val(components.r);
    greenSlider.val(components.g);
    blueSlider.val(components.b);

    field.value = components.r + "," + components.g + "," + components.b;
    delayPostColor(field.name, components);
  });

  $("#form").append(template);
}

function addColorFieldPalette(field) {
  var template = $("#colorPaletteTemplate").clone();

  var buttons = template.find(".btn-color");

  var label = template.find(".control-label");
  label.text(field.label);

  buttons.each(function(index, button) {
    $(button).click(function() {
      var rgb = $(this).css('backgroundColor');
      var components = rgbToComponents(rgb);

      field.value = components.r + "," + components.g + "," + components.b;
      postColor(field.name, components);

      ignoreColorChange = true;
      var id = "#input-" + field.name;
      $(id).minicolors("value", "rgb(" + field.value + ")");
      $(id + "-red").val(components.r);
      $(id + "-green").val(components.g);
      $(id + "-blue").val(components.b);
      $(id + "-red-slider").val(components.r);
      $(id + "-green-slider").val(components.g);
      $(id + "-blue-slider").val(components.b);
      ignoreColorChange = false;
    });
  });

  $("#form").append(template);
}

function addSectionField(field) {
  var template = $("#sectionTemplate").clone();

  template.attr("id", "form-group-section-" + field.name);
  template.attr("data-field-type", field.type);

  $("#form").append(template);
}

function updateFieldValue(name, value) {
  var group = $("#form-group-" + name);

  var type = group.attr("data-field-type");

  if (type == "Number") {
    var input = group.find(".form-control");
    input.val(value);
  } else if (type == "Boolean") {
    var btnOn = group.find("#btnOn" + name);
    var btnOff = group.find("#btnOff" + name);

    btnOn.attr("class", value ? "btn btn-primary" : "btn btn-default");
    btnOff.attr("class", !value ? "btn btn-primary" : "btn btn-default");

  } else if (type == "Select") {
    var select = group.find(".form-control");
    select.val(value);
    if (name == "pattern") {
      updateDisplayedFields(value);
    }
  } else if (type == "Color") {
    var input = group.find(".form-control");
    input.val("rgb(" + value + ")");
  }
}

function handleSetting(config) {

  for (let [key, value] of Object.entries(config.value)) {
    if (key.includes("Support") && value == true) {
      $("#" + key + "Entry").removeClass("hidden");
    } else {
      $("#input-" + key).val(value);
    }
  }

  $("#btnOnmqtt").attr("class", config.value.mqttEnabled ? "btn btn-primary" : "btn btn-default");
  $("#btnOffmqtt").attr("class", !config.value.mqttEnabled ? "btn btn-primary" : "btn btn-default");

  $("#btnOnmqtt").click(function() {
    $("#btnOnmqtt").attr("class", "btn btn-primary");
    $("#btnOffmqtt").attr("class", "btn btn-default");
    $("#input-mqttEnabled").val("1");
  });
  $("#btnOffmqtt").click(function() {
    $("#btnOnmqtt").attr("class", "btn btn-default");
    $("#btnOffmqtt").attr("class", "btn btn-primary");
    $("#input-mqttEnabled").val("0");
  });

}

function setBooleanFieldValue(field, btnOn, btnOff, value) {
  field.value = value;

  btnOn.attr("class", field.value ? "btn btn-primary" : "btn btn-default");
  btnOff.attr("class", !field.value ? "btn btn-primary" : "btn btn-default");

  postValue(field.name, field.value);
}

function postValue(name, value) {
  $("#status").html("Setting " + name + ": " + value + ", please wait...");

  var body = { name: name, value: value };

  $.post(urlBase + name + "?value=" + value, body, function(data) {
    if (data.name != null) {
      $("#status").html("Set " + name + ": " + data.name);
    } else {
      $("#status").html("Set " + name + ": " + data);
    }
  });
}

function delayPostValue(name, value) {
  clearTimeout(postValueTimer);
  postValueTimer = setTimeout(function() {
    postValue(name, value);
  }, 300);
}

function postColor(name, value) {
  $("#status").html("Setting " + name + ": " + value.r + "," + value.g + "," + value.b + ", please wait...");

  var body = { name: name, r: value.r, g: value.g, b: value.b };

  $.post(urlBase + name + "?r=" + value.r + "&g=" + value.g + "&b=" + value.b, body, function(data) {
    $("#status").html("Set " + name + ": " + data);
  })
  .fail(function(textStatus, errorThrown) { $("#status").html("Fail: " + textStatus + " " + errorThrown); });
}

function delayPostColor(name, value) {
  clearTimeout(postColorTimer);
  postColorTimer = setTimeout(function() {
    postColor(name, value);
  }, 300);
}

function componentToHex(c) {
  var hex = c.toString(16);
  return hex.length == 1 ? "0" + hex : hex;
}

function rgbToHex(r, g, b) {
  return "#" + componentToHex(r) + componentToHex(g) + componentToHex(b);
}

function rgbToComponents(rgb) {
  var components = {};

  rgb = rgb.match(/^rgb\((\d+),\s*(\d+),\s*(\d+)\)$/);
  components.r = parseInt(rgb[1]);
  components.g = parseInt(rgb[2]);
  components.b = parseInt(rgb[3]);

  return components;
}

function updateDisplayedFields(pattern = null) {

  // default value
  var pattern_index = patternData.value;

  if (pattern != null) {
    pattern_index = pattern;
  }

  this_pattern = patternData.options[pattern_index];

  if (this_pattern.show_palette == false) {
    $("#form-group-palette").addClass("hidden");
  } else {
    $("#form-group-palette").removeClass("hidden");
  }

  if (this_pattern.show_speed == false) {
    $("#form-group-speed").addClass("hidden");
  } else {
    $("#form-group-speed").removeClass("hidden");
  }

  if (this_pattern.show_color_picker == false) {
    $("#form-group-section-solidColor").addClass("hidden");
    $("#colorPaletteTemplate").addClass("hidden");
    $("#form-group-solidColor").addClass("hidden");
  } else {
    $("#form-group-section-solidColor").removeClass("hidden");
    $("#colorPaletteTemplate").removeClass("hidden");
    $("#form-group-solidColor").removeClass("hidden");
  }

  if (this_pattern.show_cooling_sparking == false) {
    $("#form-group-section-fire").addClass("hidden");
    $("#form-group-cooling").addClass("hidden");
    $("#form-group-sparking").addClass("hidden");
  } else {
    $("#form-group-section-fire").removeClass("hidden");
    $("#form-group-cooling").removeClass("hidden");
    $("#form-group-sparking").removeClass("hidden");
  }

  if (this_pattern.show_twinkle == false) {
    $("#form-group-section-twinkles").addClass("hidden");
    $("#form-group-twinkleSpeed").addClass("hidden");
    $("#form-group-twinkleDensity").addClass("hidden");
  } else {
    $("#form-group-section-twinkles").removeClass("hidden");
    $("#form-group-twinkleSpeed").removeClass("hidden");
    $("#form-group-twinkleDensity").removeClass("hidden");
  }

}